ocs_ci/utility/utils.py

```diff
@@ -1302,25 +1302,104 @@ def get_az_count():
+            tries=tries,
+            delay=delay,
+            backoff=1
+        )(ceph_health_check_base)(namespace)
+
+    degraded_pgs = 0
+    undersized_pgs = 0
+    last_decrease = time.time()
+    while tries:
+        decrease_time = None
+        degraded_pgs_diffs = []
+        degraded_pgs_diff_times = []
+        try:
+            return retry(
+                (CommandFailed),
+                tries=tries,
+                delay=delay,
+                backoff=1
+            )(ceph_health_check_base)(namespace)
+        except CephHealthException as ex:
+            str_ex = str(ex)
+            degraded_pgs_regexp = r"(\d+)(?=\s*pgs degraded)"
+            degraded_pgs_found = re.findall(degraded_pgs_regexp, str_ex)
+            if degraded_pgs_found:
+                degraded_pgs_new = int(degraded_pgs_found[0])
+                if degraded_pgs == 0 or degraded_pgs_new < degraded_pgs:
+                    decrease_time = time.time()
+                    degraded_pgs_diff_times.append(decrease_time - last_decrease)
+                    degraded_pgs_diffs.append(degraded_pgs - degraded_pgs_new)
+                    degraded_pgs = degraded_pgs_new
+                    degraded_pgs_diff_len = len(degraded_pgs_diffs)
+                    average_pgs_diff = (
+                        sum(degraded_pgs_diff_times) / degraded_pgs_diff_len
+                    )
+                    average_pgs_time_diff = (
+                        sum(degraded_pgs_diff_times) / degraded_pgs_diff_len
+                    )
+                    log.info(
+                        f"Average decrease of degraded PGs is {average_pgs_diff}, "
+                        f"average decrease time is {average_pgs_time_diff}sec."
+                    )
+            undersized_pgs_regexp = r"(\d+)(?=\s*pgs undersized)"
+            undersized_pgs_found = re.findall(undersized_pgs_regexp, str_ex)
+            if undersized_pgs_found:
+                undersized_pgs_new = int(undersized_pgs_found[0])
+                if undersized_pgs == 0 or undersized_pgs_new < undersized_pgs:
+                    decrease_time = time.time()
+                    undersized_pgs = undersized_pgs_new
+            if decrease_time:
+                last_decrease = decrease_time
+            last_time_change_diff = time.time() - last_decrease
+            if last_time_change_diff < max_timeout_for_pg:
+                if degraded_pgs or undersized_pgs:
+                    log.warn(
+                        f"Waiting for clean PGs. Degraded: {degraded_pgs} "
+                        f"Undersized: {undersized_pgs}."
+                    )
+                    log.warn(ex)
+                    log.info(
+                        f"Waiting for {delay}sec for next attempt. Timeout "
+                        f"remaining: {max_timeout_for_pg - last_time_change_diff}"
+                    )
+                    time.sleep(delay)
+                else:
```

Comment:
I am not able to visualize flow here, if we don't have any degraded or undersized PGs ( I assume pgs are  clean) why we have to sleep in this block ? why can't just loop back and wait in the ```retry``` of try block above? because if its already clean then we don't want to sleep here for ```delay``` period of time.  am I missing something ?

Reply:
If we got into this block we had to hit CephHealthException exception, that means that we had some issue with Ceph health.

We have two possibilities here:
1) degraded_pgs or undersized_pgs - so we gather those data from exception and we had degraded or undersized PGs.  The log message above this sleep print's that we are waiting for another time to see that it reduces if we didn't reach max timeout.
2) Ceph health error was caused by something else than degraded/undersized PGs.  Then we decrease number of tries below and continue in next cycle in loop. (As intelligent wait now works only for degraded/undersized PGs) but this still simulate retry as it was in non intelligent wait.

The block else with raise an exception is when we reach the timeout.  But I've just realized that we don't raise an exception in case we reach maximum re-tries, so I will change the logic here a bit to raise an exception when we move out of while loop which will solve the issue.

Hope it's clear here @shylesh ?

Reply:
got it. thanks